bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -49,13 +50,58 @@ export stream_to_column = (col, new_col, rollover) ->
   return concat_typed_arrays(col, tmp)
 
 # exported for testing
-export patch_to_column = (col, patch) ->
-  for i in [0...patch.length]
-    [ind, value] = patch[i]
-    col[ind] = value
+export slice = (ind, length) ->
+  if isObject(ind)
+    return [ind.start ? 0, ind.stop ? length, ind.step ? 1]
+  return [start, stop, step] = [ind, ind+1, 1]
 
+# exported for testing
+export patch_to_column = (col, patch, shapes) ->
+  patched = new Set()
+  patched_range = false
+
+  for [ind, value] in patch
+
+    # make the single index case look like the length-3 multi-index case
+    if not isArray(ind)
 
-# Datasource where the data is defined column-wise, i.e. each key in the
+      if Number.isInteger(ind)
```

Comment:
Not supported on IE. Also, does this have to be an integer check? If you are just differentiating between a number and something else you should use `isNumber` from `core/util/array`.